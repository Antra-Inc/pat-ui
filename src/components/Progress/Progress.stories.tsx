import React, { useState, useEffect } from 'react';
import Progress from './Progress';

export default {
  title: 'Progress',
  component: Progress,
};

export const DefaultProgress = () => {
  // linear progress bar width
  const [width, setWidth] = useState<number>(0);

  // set interval to simulate progress
  useEffect(() => {
    const interval = setInterval(() => {
      setWidth((currWidth) => {
        // reset after reaching 100
        if (currWidth === 100) {
          return 0;
        }

        // create random value
        const diff = Math.random() * 10;

        // return whichever is smaller
        return Math.min(currWidth + diff, 100);
      });
    }, 400);

    return () => {
      clearInterval(interval);
    };
  }, []);

  return <Progress pgValue={width} />;
};

export const DiffTypeProgress = () => {
  // linear progress bar width
  const [width, setWidth] = useState<number>(0);

  // set interval to simulate progress
  useEffect(() => {
    const interval = setInterval(() => {
      setWidth((currWidth) => {
        // reset after reaching 100
        if (currWidth === 100) {
          return 0;
        }

        // create random value
        const diff = Math.random() * 10;

        // return whichever is smaller
        return Math.min(currWidth + diff, 100);
      });
    }, 400);

    return () => {
      clearInterval(interval);
    };
  }, []);

  return (
    <>
      <h1>Linear</h1>
      <br />
      <Progress pgValue={width} />
      <br />

      <h1>Circular</h1>
      <br />
      <Progress pgType="circular" pgValue={width} />
      <br />
    </>
  );
};
