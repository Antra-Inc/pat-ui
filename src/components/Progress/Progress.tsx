import React, { FC } from 'react';
import { classNames } from '../../utils/classNames';

// progress type
export type ProgressType = 'circular' | 'linear';
// progress color
export type ProgressColor =
  | 'primary'
  | 'secondary'
  | 'success'
  | 'info'
  | 'warning'
  | 'danger';
// progress size
export type ProgressSize = 'xs' | 'sm' | 'lg' | 'xl';

export interface IRingProps {
  height: number;
  width: number;
  cx: number;
  cy: number;
  radius: number;
  strokeWidth: number;
}

// props type
export interface IProgressProps {
  /** set class name */
  className?: string;
  /** set progress type: circular or linear */
  pgType?: ProgressType;
  /** set message size */
  pgSize?: ProgressSize;
  /** set progress color */
  pgColor?: ProgressColor;
  /** show or hide progress text */
  showPercentage?: boolean;
  /** set progress value from 0 to 100 (%) */
  pgValue: number;
}

/**
 * A progress component shows the process of a task
 *
 * ```js
 * import {Progress} from 'pat-ui'
 * ```
 */
export const Progress: FC<IProgressProps> = (props) => {
  // props destructure
  const { className, pgType, pgSize, pgColor, showPercentage, pgValue } = props;

  // manage class names with different type
  let styleClasses;
  if (pgType === 'linear') {
    styleClasses = classNames('pg-linear', {
      [`pg-linear-${pgSize}`]: !!pgSize,
      [`pg-linear-${pgColor}`]: !!pgColor,
    });
  } else if (pgType === 'circular') {
    styleClasses = classNames('pg-circular', {
      [`pg-circular-${pgSize}`]: !!pgSize,
      [`pg-circular-${pgColor}`]: !!pgColor,
    });
  }
  // add user's custom class
  if (className) {
    styleClasses += ' ' + className;
  }

  // render different component based on type
  let renderComponent: JSX.Element | undefined;
  if (pgType === 'linear') {
    renderComponent = (
      <div style={{ display: 'flex' }}>
        <div
          style={{ width: showPercentage ? '95%' : '100%' }}
          className={styleClasses}
        >
          <div style={{ width: `${pgValue}%` }} className="pg-linear-bar"></div>
        </div>
        {showPercentage ? (
          <div className="pg-linear-text">{`${Math.floor(pgValue)}%`}</div>
        ) : null}
      </div>
    );
  } else if (pgType === 'circular') {
    let ringProps: IRingProps;

    switch (pgSize) {
      case 'xs':
        ringProps = {
          height: 25,
          width: 25,
          cx: 12.5,
          cy: 12.5,
          radius: 10,
          strokeWidth: 2.5,
        };
        break;

      case 'sm':
        ringProps = {
          height: 50,
          width: 50,
          cx: 25,
          cy: 25,
          radius: 20,
          strokeWidth: 5,
        };
        break;

      case 'lg':
        ringProps = {
          height: 100,
          width: 100,
          cx: 50,
          cy: 50,
          radius: 40,
          strokeWidth: 10,
        };
        break;

      case 'xl':
        ringProps = {
          height: 150,
          width: 150,
          cx: 75,
          cy: 75,
          radius: 60,
          strokeWidth: 15,
        };
        break;

      default:
        ringProps = {
          height: 75,
          width: 75,
          cx: 37.5,
          cy: 37.5,
          radius: 30,
          strokeWidth: 7.5,
        };
        break;
    }

    const circumference = ringProps.radius * 2 * Math.PI;
    const strokeDashoffset = circumference - (pgValue / 100) * circumference;

    renderComponent = (
      <div className={styleClasses}>
        <svg
          className="backgroud"
          height={ringProps.height}
          width={ringProps.width}
        >
          <circle
            className="ring"
            strokeWidth={ringProps.strokeWidth}
            strokeLinecap="round"
            fill="transparent"
            r={ringProps.radius}
            cx={ringProps.cx}
            cy={ringProps.cy}
          />
        </svg>
        <svg
          className="progress"
          height={ringProps.height}
          width={ringProps.width}
        >
          <circle
            className="progress-ring"
            style={{ strokeDashoffset }}
            strokeWidth={ringProps.strokeWidth}
            strokeLinecap="round"
            strokeDasharray={circumference + ' ' + circumference}
            fill="transparent"
            r={ringProps.radius}
            cx={ringProps.cx}
            cy={ringProps.cy}
          />
        </svg>
        {showPercentage ? (
          <div className="pg-circular-text">{`${Math.floor(pgValue)}%`}</div>
        ) : null}
      </div>
    );
  }

  return <>{renderComponent}</>;
};

Progress.defaultProps = {
  pgType: 'linear',
};

export default Progress;
