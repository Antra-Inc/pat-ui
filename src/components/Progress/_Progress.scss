.linear {
  width: 100%;
  background-color: lightgrey;

  .bar {
    width: 0%;
    height: 30px;
    background-color: lightgreen;
  }
}
