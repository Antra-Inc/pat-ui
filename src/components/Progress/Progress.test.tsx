import React from 'react';

import { render } from '@testing-library/react';
import Progress, { IProgressProps } from './Progress';

describe('Progress', () => {
  // snapshot testing
  it('should match snapshot', () => {
    const { asFragment } = render(<Progress pgValue={0}></Progress>);
    expect(asFragment()).toMatchSnapshot();
  });

  // default progress (linear)
  it('should render default progress, which is linear progress', () => {
    const wrapper = render(<Progress pgValue={0}></Progress>);
  });

  // circular progress
  // it('should render circular progress', () => {});

  // render progress based on props
  // it('should render correct progress based on different props', () => {});
});
