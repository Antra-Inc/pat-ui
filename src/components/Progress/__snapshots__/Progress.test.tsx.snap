// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress should match snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: flex;"
  >
    <div
      class="pg-linear"
      style="width: 100%;"
    >
      <div
        class="pg-linear-bar"
        style="width: 0%;"
      />
    </div>
  </div>
</DocumentFragment>
`;
